Update Tests and Add Demos for Enforced Updating (android) (by @appknox) (#3462)
OWASP/mastg: tests-beta/android/MASVS-CODE/MASTG-TEST-0x36.md

---
title: Enforcing Mandatory In-App Updates
platform: android
id: MASTG-TEST-0x36
type: [static]
weakness: MASWE-0075
profiles: [L2]
---

## Overview

> The goal of this test is to verify whether the application enforces mandatory updates, preventing users from accessing the app until the latest version has been successfully downloaded and installed. A mandatory update can typically be achieved by using the [Google Play Core In-App Update API](https://developer.android.com/guide/playcore/in-app-updates/kotlin-java) and invoking `startUpdateFlowForResult` with an Immediate update type option `AppUpdateType.IMMEDIATE` or value `1`.

Reviewer: ```suggestion
The goal of this test is to verify whether the application enforces mandatory updates and prevents users from accessing the app until the latest version has been successfully downloaded and installed. 

A mandatory update can typically be achieved by using the [Google Play Core In-App Update API](https://developer.android.com/guide/playcore/in-app-updates/kotlin-java) and invoking `startUpdateFlowForResult` with an Immediate update type option `AppUpdateType.IMMEDIATE` or value `1`.
```

> The goal of this test is to verify whether the application enforces updates in a way that blocks usage until the latest version is installed. This is typically achieved using the [Google Play Core In-App Update API](https://developer.android.com/guide/playcore/in-app-updates/kotlin-java) by invoking `startUpdateFlowForResult` with an Immediate update type option `AppUpdateType.IMMEDIATE` or value `1`. This configuration initiates a non-cancellable, blocking update flow. The test should involve launching the app when an update is available and verifying that access to the app's functionality remains restricted until the update has been successfully downloaded and installed.

Author: The app registers an ActivityResultLauncher to handle the update flow result. In the `MainActivity.kt` (lines 36-43) If the update fails or the user cancels it (result.resultCode != RESULT_OK) , the app logs the event and immediately re-triggers mastgTest.checkForUpdate(...), causing the update prompt to appear again.

## Steps

> 1. Run a static analysis tool such as @MASTG-TOOL-0110 on codebase for usages of the calls to the Play Core in-app update API, specifically `startUpdateFlowForResult`, that are configured with the integer value `1`.

Reviewer: ```suggestion
1. Run a static analysis tool such as @MASTG-TOOL-0110 on codebase for usages of the calls to the Play Core in-app update API, specifically `startUpdateFlowForResult`, that are configured with the integer value `1` (`AppUpdateType.IMMEDIATE`).
```

## Observation

The output should contain the locations where `startUpdateFlowForResult` with `AppUpdateOptions.newBuilder(1).build()` is called.

## Evaluation

> The test fails if the app does not enforce updates and still allows users to skip or ignore them.

Reviewer: Could you please indicate what should look for in the code based on the above observation in order to evaluate this?

Author: The app registers an ActivityResultLauncher to handle the update flow result. In the MainActivity.kt (lines 36-43) if the update fails or the user cancels it (result.resultCode != RESULT_OK), the app logs the event and immediately re-triggers mastgTest.checkForUpdate(...), causing the update prompt to appear again.

---
title: Enforcing Mandatory In-App Updates
platform: android
id: MASTG-TEST-0x36
type: [static]
weakness: MASWE-0075
profiles: [L2]
knowledge: [MASTG-KNOW-0023]
---

## Overview

If the app does not properly enforce mandatory updates, users may continue running outdated versions that contain known security vulnerabilities or deprecated functionality. This can lead to exploitation of patched vulnerabilities, data leakage, or incompatibility with backend security requirements.

> On Android, mandatory updates can be implemented using the [Google Play Core In-App Update API](https://developer.android.com/guide/playcore/in-app-updates/kotlin-java) by invoking `startUpdateFlowForResult` with `AppUpdateType.IMMEDIATE` (value `1`). However, simply calling `startUpdateFlowForResult` is insufficient; users can dismiss the update dialog or background the app before the update completes. Without additional enforcement in the `onResume` lifecycle method to detect `UPDATE_AVAILABLE` or `DEVELOPER_TRIGGERED_UPDATE_IN_PROGRESS` states, users can bypass the update entirely.

Reviewer 2: - let's describe why the app should use  `AppUpdateType.IMMEDIATE` instead of `AppUpdateType.FLEXIBLE`. It can be in a KNOW page but we definitely should cover it. 
- make sure to link the API to the relevant Android docs. See my suggestions
- Not sure if we need to write constant values in the test. Tests should be generic - "(value `1`)". We can write about it in the demo

```suggestion
On Android, mandatory updates can be implemented using the [Google Play Core In-App Update API](https://developer.android.com/guide/playcore/in-app-updates/kotlin-java) by invoking [`startUpdateFlowForResult`](https://developer.android.com/reference/com/google/android/play/core/appupdate/AppUpdateManager#startUpdateFlowForResult(com.google.android.play.core.appupdate.AppUpdateInfo,%20android.app.Activity,%20com.google.android.play.core.appupdate.AppUpdateOptions,%20int)) with [`AppUpdateType.IMMEDIATE`](https://developer.android.com/reference/com/google/android/play/core/install/model/AppUpdateType.html#IMMEDIATE). However, simply calling `startUpdateFlowForResult` is insufficient; users can dismiss the update dialog or background the app before the update completes. Without additional enforcement in the `onResume` lifecycle method to detect `UPDATE_AVAILABLE` or `DEVELOPER_TRIGGERED_UPDATE_IN_PROGRESS` states, users can bypass the update entirely.
```
